db/db_versioned.go

```diff
@@ -0,0 +1,151 @@
+	//
+	// Here's an example of a versioned DB which has 3 buckets:
+	// 1. "mta" -- regular bucket storing metadata, key is not versioned
+	// 2. "act" -- versioned namespace, key length = 20
+	// 3. "stk" -- versioned namespace, key length = 8
+	KvVersioned interface {
+		KVStore
+
+		// Version returns the key's most recent version
+		Version(string, []byte) (uint64, error)
+
+		// SetVersion sets the version before calling Put()
+		SetVersion(uint64) KvVersioned
+
+		// CreateVersionedNamespace creates a namespace to store versioned keys
+		CreateVersionedNamespace(string, uint32) error
+	}
+
+	// BoltDBVersioned is KvVersioned implementation based on bolt DB
+	BoltDBVersioned struct {
+		*BoltDB
+		version      uint64          // version for Get/Put()
+		versioned    map[string]int  // buckets for versioned keys
+		nonversioned map[string]bool // buckets for non-versioned keys
+	}
+)
+
+// Option sets an option
+type Option func(b *BoltDBVersioned)
+
+// NonversionedNamespaceOption sets non-versioned namespace
+func NonversionedNamespaceOption(ns ...string) Option {
+	return func(b *BoltDBVersioned) {
+		for _, v := range ns {
+			b.nonversioned[v] = true
+		}
+	}
+}
+
+// VersionedNamespaceOption sets versioned namespace
+func VersionedNamespaceOption(ns string, n int) Option {
+	return func(b *BoltDBVersioned) {
+		b.versioned[ns] = n
+	}
+}
+
+// NewBoltDBVersioned instantiates an BoltDB with implements KvVersioned
+func NewBoltDBVersioned(cfg Config, opts ...Option) *BoltDBVersioned {
+	b := &BoltDBVersioned{
+		BoltDB:       NewBoltDB(cfg),
+		versioned:    make(map[string]int),
+		nonversioned: make(map[string]bool),
+	}
+	for _, opt := range opts {
+		opt(b)
+	}
+	return b
+}
+
+// Start starts the DB
+func (b *BoltDBVersioned) Start(ctx context.Context) error {
+	if err := b.BoltDB.Start(ctx); err != nil {
+		return err
+	}
+	// TODO: check non-versioned and versioned namespace
+	return nil
+}
+
+// Put writes a <key, value> record
+func (b *BoltDBVersioned) Put(ns string, key, value []byte) error {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+	// TODO: implement Put
+	return nil
+}
+
+// Get retrieves the most recent version
+func (b *BoltDBVersioned) Get(ns string, key []byte) ([]byte, error) {
+	if !b.IsReady() {
+		return nil, ErrDBNotStarted
+	}
+	// TODO: implement Get
+	return nil, nil
+}
+
+// Delete deletes a record,if key is nil,this will delete the whole bucket
+func (b *BoltDBVersioned) Delete(ns string, key []byte) error {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+	// TODO: implement Delete
+	return nil
+}
+
+// Version returns the key's most recent version
+func (b *BoltDBVersioned) Version(ns string, key []byte) (uint64, error) {
+	if !b.IsReady() {
+		return 0, ErrDBNotStarted
+	}
+	// TODO: implement Version
+	return 0, nil
+}
+
+// SetVersion sets the version which a Get()/Put() wants to access
+func (b *BoltDBVersioned) SetVersion(v uint64) KvVersioned {
```

Comment:
the v means `blockHeight` in production?

Reply:
yes, that's correct